src/lib/ares_search.c

```diff
@@ -33,57 +33,91 @@
-  size_t          next_domain;  /* next search domain to try */
-  ares_bool_t     trying_as_is; /* current query is for name as-is */
-  size_t          timeouts;     /* number of timeouts we saw for this request */
-  ares_bool_t ever_got_nodata; /* did we ever get ARES_ENODATA along the way? */
+  /* State tracking progress through the search query */
+  int             status_as_is;    /* error status from trying as-is */
+  size_t          next_domain;     /* next search domain to try */
+  ares_bool_t     trying_as_is;    /* current query is for name as-is */
+  size_t          timeouts;        /* number of timeouts we saw for this request */
+  ares_bool_t     ever_got_nodata; /* did we ever get ARES_ENODATA along the way? */
 };
 
 static void search_callback(void *arg, int status, int timeouts,
                             unsigned char *abuf, int alen);
 static void end_squery(struct search_query *squery, ares_status_t status,
                        unsigned char *abuf, size_t alen);
 
-static void ares_search_int(ares_channel_t *channel, const char *name,
-                            int dnsclass, int type, ares_callback callback,
-                            void *arg)
+static void ares_search_int(ares_channel_t *channel, ares_dns_record_t *dnsrec,
+                            ares_callback callback, void *arg)
 {
   struct search_query *squery;
-  char                *s;
+  const char          *name;
+  char                *s = NULL;
+  char                *qname;
+  unsigned char       *buf = NULL;
+  size_t               buflen;
   const char          *p;
   ares_status_t        status;
   size_t               ndots;
 
+  /* Extract the name for the search. Note that searches are only supported for
+   * DNS records containing a single query.
+   */
+  if (ares_dns_record_query_cnt(dnsrec) != 1) {
+    callback(arg, ARES_EBADQUERY, 0, NULL, 0);
+    return;
+  }
+  ares_dns_record_query_get(dnsrec, 0, &name, NULL, NULL);
+
   /* Per RFC 7686, reject queries for ".onion" domain names with NXDOMAIN. */
   if (ares__is_onion_domain(name)) {
     callback(arg, ARES_ENOTFOUND, 0, NULL, 0);
     return;
   }
 
   /* If name only yields one domain to search, then we don't have
-   * to keep extra state, so just do an ares_query().
+   * to keep extra state, so just do an ares_send().
    */
   status = ares__single_domain(channel, name, &s);
   if (status != ARES_SUCCESS) {
     callback(arg, (int)status, 0, NULL, 0);
     return;
-  }
-  if (s) {
-    ares_query(channel, s, dnsclass, type, callback, arg);
+  } else if (s != NULL) {
+    /* We only have a single domain to search, so do it here. The domain to
+     * search may be different to the name passed in, so temporarily
+     * overwrite it before encoding the DNS record.
+     *
+     * TODO: Temporarily overwriting the name should be a helper function.
+     *       Modifying the ares_dns_record_t structure internals should not be
+     *       done here.
+     */
```

Comment:
I have moved this into a helper function called `ares_dns_write_query_altname`, to temporarily overwrite the query name before calling `ares_dns_write`.

Note that I haven't added a function to modify the question names on a DNS record more generally; I think the memory ownership is harder to reason about (because `ares_dns_record_query_get` returns pointers to the underlying data), and we only need this specific case.